src/core/cgroup.c

```diff
@@ -1853,7 +1909,7 @@ static int on_cgroup_inotify_event(sd_event_source *s, int fd, uint32_t revents,
                                  * this here safely. */
                                 continue;
 
-                        (void) unit_notify_cgroup_empty(u);
+                        unit_add_to_cgroup_empty_queue(u);
```

Comment:
Hmm, I'm probably missing something, but I'll ask anyway.
Why can't the unit be added to `gc_queue` directly?

Reply:
different unit types want to react differently on their cgroup running empty. For example, service units might want to go through ExecStopPost= or so, and do other stuff. Hence each unit type can install a handler that is called whenever its cgroup runs empty, and it can do there whatever it likes. Now, we need to make sure that if a service dies, we always process the last SIGCHLD we might get for the service before the cgroup empty event (simply because the SIGCHLD is infinitely more interesting, as it contains the process exit status and so on, which might mean we need to enter a different state for the unit). And this "cgroup_empty" queue is here to make sure this happens in the proper order: whenever we notice a cgroup runs empty, we add the unit to the queue, schedule a "defer" event for dispatching the queue, but do so at a lower event loop priority than the SIGCHLD handling, so that the SIGCHLD handling is processed first.

The GC queue otoh is where we add units that we think might be worthy of unloading from memory. It's a very different thing really. Units are unloaded from memory if they are dead, have no processes associated with them anymore, no client refs and so on. 

Hope that makes sense?

Reply:
Ah, `on_cgroup_empty_event` calls `UNIT_VTABLE(u)->notify_cgroup_empty(u)`. I missed this, it was just after outside of the range shown by the diff. Thanks.